File: src/maths.rs
use std::f64::EPSILON;

fn sqrt(num: f64) -> f64 {
    let mut y: f64 = num;
    let x2: f64 = y * 0.5;
    let i: *mut u64 = unsafe {
        &mut *(&mut y as *mut f64 as *mut u64)
    };

    unsafe { *i = 0x5fe6eb50c7b537a9 - (*i >> 1) };
    y = unsafe {
        *(i as *mut f64)
    };
    for _ in 0..6 {
        y = y * (1.5 - (x2 * y * y));
    }

    1.0 / y
}

pub struct Polynomial {
    pub a: f64,
    pub b: f64,
    pub c: f64,
}

pub struct Complex {
    pub a: f64,
    pub b: f64,
}

impl Polynomial {

    pub fn new() -> Polynomial {
        Polynomial { a: 0.0, b: 0.0, c: 0.0}
    }
    
    pub fn delta(&self) -> f64 {
        self.b * self.b - 4.0 * self.a * self.c 
    }

    pub fn degree(&self) -> i64 {
        let degree: i64;

        if self.a > EPSILON || self.a < -EPSILON {
            degree = 2;
        } else if self.b > EPSILON || self.b < -EPSILON {
            degree = 1;
        } else {
            degree = 0;
        }

        degree
    }

    fn solve_1(&self) -> Vec<Complex> {
        let mut res = Vec::<Complex>::new();

        res.push(Complex {
            a: (- self.c) / self.b,
            b: 0.0,
        });
        res
    }

    fn solve_2(&self) -> Vec<Complex> {
        let delta = self.delta();
        let mut res = Vec::<Complex>::new();

        if delta > EPSILON // Two solutions
        {
            res.push(Complex {
                a: (- self.b - sqrt(delta)) / (2.0 * self.a),
                b: 0.0,
            });
            res.push(Complex {
                a: (- self.b + sqrt(delta)) / (2.0 * self.a),
                b: 0.0,
            });
        }
        else if delta < -EPSILON  // Complex solutions
        {
            res.push(Complex {
                a: (- self.b) / (2.0 * self.a),
                b: (- sqrt(- delta)) / (2.0 * self.a),
            });
            res.push(Complex {
                a: (- self.b) / (2.0 * self.a),
                b: (sqrt(- delta)) / (2.0 * self.a),
            });
        }
        else // One solution (== 0)
        {
            res.push(Complex {
                a: (- self.b) / (2.0 * self.a),
                b: 0.0,
            });
        }
        res
    }

    pub fn solve(&self) -> Vec<Complex> {
        let degree = self.degree();

        match degree {
            2 => self.solve_2(),
            1 => self.solve_1(),
            _ => { Vec::<Complex>::new() }
        }
    }
}

pub fn reduce(left: &mut Polynomial, right: &mut Polynomial) {
    left.a = left.a - right.a;
    left.b = left.b - right.b;
    left.c = left.c - right.c;
    right.a = 0.0;
    right.b = 0.0;
    right.c = 0.0;
}

File: lib/maths.js
const error = require('./error')

const buf = new ArrayBuffer(4)
const f32 = new Float32Array(buf)
const u32 = new Uint32Array(buf)

function sqrt(x) {
	const x2 = 0.5 * (f32[0] = x)
	u32[0] = (0x5f3759df - (u32[0] >> 1)) // What the fuck ?
	let y = f32[0]
	for (let i = 0; i < 10; i++)
		y = y * ( 1.5 - ( x2 * y * y ));   // 1st iteration
	return 1 / y
}

function polynome(equation) {
	this.a = 0
	this.b = 0
	this.c = 0
	this.delta = null

	for (const token of equation) {
		switch (token.order) {
			case 2:
				this.a += token.factor
				break
			case 1:
				this.b += token.factor
				break
			case 0:
				this.c += token.factor
				break
		}
	}

	this.getDelta = () => {
		return (this.b * this.b) - (4 * this.a * this.c)
	}

	this.solve2 = () => {
		this.delta = this.getDelta()
		let res = []

		if (this.delta == 0) {
			res.push({
				a: -this.b / 2 * this.a,
				b: 0
			})
		} else if (this.delta > 0) {
			let deltaSqrt = sqrt(this.delta)

			res.push({
				a: (-this.b - deltaSqrt) / (2 * this.a),
				b: 0
			})
			res.push({
				a: (-this.b + deltaSqrt) / (2 * this.a),
				b: 0
			})
		} else {
			let deltaSqrt = sqrt(-this.delta)

			res.push({
				a: -this.b / (2 * this.a),
				b: -deltaSqrt / (2 * this.a)
			})
			res.push({
				a: -this.b / (2 * this.a),
				b: deltaSqrt / (2 * this.a)
			})
		}
		return res
	}

	this.solve1 = () => {
		let res = []

		res.push({
			a: -this.c / this.b,
			b: 0
		})
		return res
	}

	this.solve = () => {
		let res

		if (this.a)
			res = this.solve2()
		else if (this.b)
			res = this.solve1()
		else if (this.c)
			error("Provided equation has no solution")
		else
			error("Any number is a solution")
		return res
	}
}

function getOrder(equation) {
	let order = 0;

	for (const token of equation) {
		if (token.order > order)
			order = token.order;
		if (token.order < 0)
			error(`Provided equation is not valid : One of its component has a negative degree ${token.order}`)
		if (!Number.isInteger(token.order))
			error(`Provided equation is not valid : One of its component has a non-integer degree ${token.order}`)
	}
	return order
}

module.exports = {getOrder: getOrder, polynome: polynome}
